libs/threading_base/include/hpx/threading_base/scheduler_base.hpp

```diff
@@ -276,15 +276,38 @@ namespace hpx { namespace threads { namespace policies {
             return thread_queue_init_.small_stacksize_;
         }
 
-        void set_user_polling_function(util::function_nonser<void()> const& f)
+        using PollingFunctionPtr = void (*)();
+
+        static void null_polling_function() {}
+
+        void set_mpi_polling_function(PollingFunctionPtr mpi_func)
+        {
+            polling_function_mpi_ = mpi_func;
+        }
+
+        void clear_mpi_polling_function()
+        {
+            polling_function_mpi_ = &null_polling_function;
+        }
+
+        void set_cuda_polling_function(PollingFunctionPtr cuda_func)
         {
-            user_polling_function_ = f;
+            polling_function_cuda_ = cuda_func;
         }
 
-        void user_polling_function()
+        void clear_cuda_polling_function()
         {
-            if (user_polling_function_)
-                user_polling_function_();
+            polling_function_cuda_ = &null_polling_function;
+        }
+
+        inline void custom_polling_function() const
+        {
+#if defined(HPX_HAVE_LIB_ASYNC_MPI)
+            (*polling_function_mpi_)();
```

Comment:
Do you know if there's a difference in performance compared to `if (polling_function_mpi_ != nullptr) (*polling_function_mpi_)();`?

Reply:
```
if (polling_function_mpi_ != nullptr) (*polling_function_mpi_)();
```
could be subject to a race if `polling_function_mpi` is set to `nullptr` between the test and the invoke. For this reason, I settled on a null function, or a valid function, with no `if ... then` check.